packages/zoltan2/src/problems/Zoltan2_PartitioningSolution.hpp

```diff
@@ -383,6 +383,97 @@ template <typename Adapter>
     else                   return NULL;
   }
 
+  /*! \brief returns the partition tree nodes.
+   */
+  virtual const std::vector<partitionTreeNode<part_t> > &
+  getPartitionTreeNodes() const
+  {
+    return this->algorithm_->getPartitionTreeNodes();
+  }
+
+  /*! \brief calculate if partition tree is binary.
+   */
+  virtual bool isPartitioningTreeBinary() const
+  {
+    // returns true if tree is binary
+    // TO DO - determine if this should be true if tree is in fact binary
+    // or must have been binary - for example if an MJ algorithm could have
+    // done a 3-split but all end up being 2, do we want true here?
+    if(getPartitionTreeNodes().size() == 0) {
+      throw std::logic_error(
+        "no binary tree available. Set param 'keep_partition_tree' true");
+    }
+    for(part_t n = 0; n < static_cast<part_t>(getPartitionTreeNodes().size()); ++n) {
+      if(getPartitionTreeNodes()[n].children.size() > 2) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /*! \brief build a list of all children under the indexed node
+   */
+  void recursiveCollectAllParts(part_t nodeIndex, std::vector<part_t> & parts) const {
+    // traverse the tree filling with all children
+    const partitionTreeNode<part_t> & node = getPartitionTreeNodes()[nodeIndex];
+    // note this is not part_t because the children of a node is expected to
+    // always be a short list - for rcb it will be just 2 elements
+    for(int n = 0; n < static_cast<int>(node.children.size()); ++n) {
+      part_t child = node.children[n];
+      if(child <= 0) { // this is a terminal
+        parts.push_back(-child); // terminal = -child convention
+      }
+      else { // if not terminal, -1 for convention and recursive continue
+        recursiveCollectAllParts(child-1, parts); // node+1 convention
+      }
+    }
+  }
+
+  /*! \brief get the partition tree - fill the relevant arrays
+   */
+  void getPartitionTree(part_t & numTreeVerts,
```

Comment:
Would this function be simpler if it were a call-through function to algorithm-specific implementations that accessed the needed data in its natural format (similar to the way boxAssign and pointAssign work)?   That way, the algorithms wouldn't have to convert their natural info (e.g., rcb_tree treeptr in RCB) to a specific format (partitionTreeNode) and then convert it again to the array format for getPartitionTree. Then, for example, RCB could fill the arrays directly from its rcb_tree treeptr.